libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2184,20 +2184,44 @@ void TypeChecker::endVisit(ElementaryTypeNameExpression const& _expr)
 
 void TypeChecker::endVisit(Literal const& _literal)
 {
-	bool const v050 = m_scope->sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::V050);
+	bool const v050 = m_scope->sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::V050) > 0;
 
 	if (_literal.looksLikeAddress())
 	{
-		if (_literal.passesAddressChecksum())
+		bool acceptableAddress = true;
+		if (_literal.value().length() != 42) // "0x" + 40 hex digits
+		{
+			string error =
+					"This looks like an address but is not exactly 20 bytes long. Current length is " + to_string(_literal.value().length()) +
+					" (should be 42).";
+
+			if (v050) {
+				m_errorReporter.syntaxError(_literal.location(), error);
+				acceptableAddress = false;
+			}
+			else
+				m_errorReporter.warning(error);
+		}
+		else if (!_literal.passesAddressChecksum())
+		{
+			string error =
+					"This looks like an address but has an invalid checksum. "
+					"If this is not used as an address, please prepend '00'. " +
+					(!_literal.getChecksummedAddress().empty() ? "Correct checksummed address: '" +
+																 _literal.getChecksummedAddress() + "'. " : "") +
+					"For more information please see https://solidity.readthedocs.io/en/develop/types.html#address-literals";
+
+			if (v050)
+			{
+				m_errorReporter.syntaxError(_literal.location(), error);
+				acceptableAddress = false;
+			}
+			else
+				m_errorReporter.warning(_literal.location(), error);
+		}
+
+		if (acceptableAddress)
 			_literal.annotation().type = make_shared<IntegerType>(160, IntegerType::Modifier::Address);
```

Comment:
If `annotation().type` is not assigned, we should have a fatal error (i.e. an exception) and thus I think the `acceptableAddress` variable is not required.

Reply:
In other words: If this function returns, then `annotation().type` has to be assigned some type.